nomad 1.3.0-rc.1 cgroupsv2 /dev/ strangeness

### Nomad version
Nomad v1.3.0-rc.1 (31b0a1841cb6d3c7aa64bb2162f50798ba3b3d6e)

### Operating system and Environment details
Ubuntu 22.04 Jammy Jellyfish  5.15.0-27-generic #28-Ubuntu SMP Thu Apr 14 04:55:28 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Issue
Something funky happens when using cgroups v2 where (at the very least) the ability to read /dev/zero and write to /dev/null is lost. Using a raw_exec job and `dd if=/dev/zero of=/dev/null count=1` I get `dd: failed to open '/dev/zero': Operation not permitted` fairly randomly but consistently. The original jobs are actually python/java, not that it seems to matter.

If I disable cgroupsv2 with the following, the problem disappears.
```diff --git a/client/lib/cgutil/cgutil_linux.go b/client/lib/cgutil/cgutil_linux.go
index 1333d0cc1..5f8d8fc68 100644
--- a/client/lib/cgutil/cgutil_linux.go
+++ b/client/lib/cgutil/cgutil_linux.go
@@ -18,7 +18,7 @@ import (
 // cgroups.v1
 //
 // This is a read-only value.
-var UseV2 = cgroups.IsCgroup2UnifiedMode()
+var UseV2 = false //cgroups.IsCgroup2UnifiedMode()

 // GetCgroupParent returns the mount point under the root cgroup in which Nomad
 // will create cgroups. If parent is not set, an appropriate name for the version
 ```

Output from a failed run, I'll note it seems to always work on the "first" run, only seems to be restarts under the same allocation id.
```
+0 records in
1+0 records out
512 bytes copied, 0.000455007 s, 1.1 MB/s
1+0 records in
1+0 records out
512 bytes copied, 0.000992872 s, 516 kB/s
1+0 records in
1+0 records out
512 bytes copied, 0.000578372 s, 885 kB/s
1+0 records in
1+0 records out
512 bytes copied, 0.000302934 s, 1.7 MB/s
1+0 records in
1+0 records out
512 bytes copied, 0.000195964 s, 2.6 MB/s
dd: failed to open '/dev/zero': Operation not permitted

File: /alloc/logs/env.stdout.0
busted
total 124K
lrwxrwxrwx   1 root root     7 Apr 23 02:08 bin -> usr/bin
lrwxrwxrwx   1 root root     8 Apr 23 02:08 sbin -> usr/sbin
lrwxrwxrwx   1 root root    10 Apr 23 02:08 libx32 -> usr/libx32
lrwxrwxrwx   1 root root     9 Apr 23 02:08 lib64 -> usr/lib64
lrwxrwxrwx   1 root root     9 Apr 23 02:08 lib32 -> usr/lib32
lrwxrwxrwx   1 root root     7 Apr 23 02:08 lib -> usr/lib
drwxr-xr-x   2 root root  4.0K Apr 23 02:08 srv
drwxr-xr-x   2 root root  4.0K Apr 23 02:08 media
drwxr-xr-x  14 root root  4.0K Apr 23 02:08 usr
drwxr-xr-x   6 root root  4.0K Apr 23 02:09 snap
drwx------   2 root root   16K Apr 23 02:09 lost+found
drwxr-xr-x   4 root root  4.0K May  5 01:44 boot
drwxr-xr-x   2 root root  4.0K May  5 01:49 mnt
drwxr-xr-x   5 root wheel 4.0K May  5 01:49 nfs
drwxr-xr-x 158 root root  4.0K May  5 01:50 home
drwxr-xr-x  14 root root  4.0K May  5 01:51 var
dr-xr-xr-x  13 root root     0 May  5 02:03 sys
dr-xr-xr-x 539 root root     0 May  5 02:03 proc
drwxr-xr-x  20 root root  4.0K May  5 02:03 ..
drwxr-xr-x  20 root root  4.0K May  5 02:03 .
drwxr-xr-x  19 root root  4.1K May  5 02:03 dev
drwxr-xr-x 141 root root   12K May  5 03:01 etc
drwxr-xr-x  48 root root  4.0K May  5 04:58 opt
drwx------   6 root root  4.0K May  5 05:14 root
drwxr-xr-x  39 root root  1.3K May  5 05:35 run
drwxrwxrwt  15 root root   40K May  5 05:36 tmp
crw-rw-rw- 1 root root 1, 5 May  5 02:03 /dev/zero
uid=0(root) gid=0(root) groups=0(root)
Current: =ep
Bounding set =cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read,cap_perfmon,cap_bpf,cap_checkpoint_restore
Ambient set =
Current IAB:
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)
 secure-no-ambient-raise: no (unlocked)
uid=0(root) euid=0(root)
gid=0(root)
groups=
Guessed mode: UNCERTAIN (0)
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) 0
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 3182
max locked memory           (kbytes, -l) unlimited
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1048576
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) unlimited
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited

### Job file (if appropriate)
```
job "test-devnull" {
	datacenters = ["..."]
	type = "service"

	constraint {
		attribute = "${attr.unique.hostname}"
		value = "..."
	}

       reschedule {
	    delay = "5s"
	    delay_function = "constant"
	    unlimited = true
       }

	group "group" {
		restart {
			attempts = 5
			mode = "fail"
			delay = "0s"
			interval = "30s"
		}
	
               count = 10

		task "env" {
			driver = "raw_exec"
			config {
				command = "/usr/bin/bash"
				args = [ "-c", "dd if=/dev/zero of=/dev/null count=1 || (echo \"busted\"; ls -ltrha /; ls -ltrha /dev/zero; id; capsh --print; ulimit -a; sleep 1000)"]
			}
			resources {
				memory = "10"
			}
		}

	}
}```

So I had a little time today and played with this a bit. on 1.3.2-dev

After a few runs, I was able to tease this out of the logs:

> > Jun 10 01:36:46 cd01-webdeb1 nomad[2605308]: time="2022-06-10T01:36:46Z" level=info msg="found more than one filter (2) attached to a cgroup -- removing extra filters!"
> > Jun 10 01:36:46 cd01-webdeb1 nomad[2605308]: time="2022-06-10T01:36:46Z" level=info msg="removing old filter 0 from cgroup" id=49060 name= run_count=0 runtime=0s tag=fb6cb1c301453333 type=CGroupDevice
> > Jun 10 01:36:46 cd01-webdeb1 nomad[2605308]: time="2022-06-10T01:36:46Z" level=info msg="removing old filter 1 from cgroup" id=49061 name= run_count=0 runtime=0s tag=531db05b114e9af3 type=CGroupDevice
> > Jun 10 01:36:46 cd01-webdeb1 nomad[2605308]:     2022-06-10T01:36:46.448Z [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=1a0affdc-a314-480a-d2d3-135d785524ed task=env error="failed to launch command with executor: rpc error: code = Unknown desc = failed to set v2 cgroup resources: failed to call BPF_PROG_DETACH (BPF_CGROUP_DEVICE) on old filter program: can't detach program: no such file or directory"
> > Jun 10 01:36:50 cd01-webdeb1 nomad[2605308]:     2022-06-10T01:36:50.481Z [WARN]  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=1a0affdc-a314-480a-d2d3-135d785524ed task=env @module=logmon timestamp=2022-06-10T01:36:50.480Z
> > Jun 10 01:36:50 cd01-webdeb1 nomad[2605308]:     2022-06-10T01:36:50.481Z [WARN]  client.alloc_runner.task_runner.task_hook.logmon.nomad: timed out waiting for read-side of process output pipe to close: alloc_id=1a0affdc-a314-480a-d2d3-135d785524ed task=env @module=logmon timestamp=2022-06-10T01:36:50.480Z
> 

I can get that fairly consistently. Not as frequent as the OP. Which, to me, looks like a symptom of the same problem.

So on a whim I tried this patch:
```
diff --git a/drivers/shared/executor/executor_universal_linux.go b/drivers/shared/executor/executor_universal_linux.go
index ac79b518b..91a76d0e8 100644
--- a/drivers/shared/executor/executor_universal_linux.go
+++ b/drivers/shared/executor/executor_universal_linux.go
@@ -108,6 +108,7 @@ func (e *UniversalExecutor) configureResourceContainer(pid int) error {
                path := filepath.Join("/", cgutil.GetCgroupParent(parent), scope)
                cfg.Cgroups.Path = path
                e.containment = resources.Contain(e.logger, cfg.Cgroups)
+               cfg.Cgroups.Resources.SkipDevices = true
                return e.containment.Apply(pid)

        } else {

```

Which breaks tasks consistently with a similar looking error message, albeit much earlier with:

> Time                  Type             Description
> 2022-06-10T02:00:08Z  Alloc Unhealthy  Unhealthy because of failed task
> 2022-06-10T02:00:08Z  Not Restarting   Error was unrecoverable
> 2022-06-10T02:00:08Z  Driver Failure   failed to launch command with executor: rpc error: code = Unknown desc = failed to start command path="/usr/bin/bash" --- args=["/usr/bin/bash" "-c" "dd if=/dev/zero of=/dev/null count=1  <none>  (echo \"busted\"; ls -ltrha /; ls -ltrha /dev/zero; id; capsh --print; ulimit -a; cat /proc/self/cgroup; sleep 1000)"]: open /dev/null: operation not permitted
> 2022-06-10T02:00:07Z  Task Setup       Building Task Directory
> 2022-06-10T02:00:07Z  Received         Task received by client

Kind of all leads me back to https://github.com/hashicorp/nomad/pull/12875 ?

In random googling it also sounds like when launching nomad with systemd `Delegate=yes` might be wanted per https://systemd.io/CGROUP_DELEGATION/#some-donts. I tried with that with no avail. Although the picture in my head of systemd unit cgroups and nomad cgroupv2 vs v1 paths is very fuzzy.

This is still easily reproducible on a vanilla arch linux install running the nomad binary as root from the command line, and ubuntu 22.04.

--- a/.changelog/17535.txt
+++ b/.changelog/17535.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+cgroups: Fixed a bug removing all DevicesSets when alloc is created/removed
+```

--- a/client/lib/cgutil/cpuset_manager_v2.go
+++ b/client/lib/cgutil/cpuset_manager_v2.go
@@ -330,7 +330,8 @@ func (c *cpusetManagerV2) write(id identity, set cpuset.CPUSet) {
 
 	// set the cpuset value for the cgroup
 	if err = m.Set(&configs.Resources{
-		CpusetCpus: set.String(),
+		CpusetCpus:  set.String(),
+		SkipDevices: true,
 	}); err != nil {
 		c.logger.Error("failed to set cgroup", "path", path, "error", err)
 		return

--- a/e2e/isolation/devices_test.go
+++ b/e2e/isolation/devices_test.go
@@ -1,0 +1,54 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package isolation
+
+import (
+	"testing"
+
+	"github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/shoenig/test/must"
+)
+
+func TestCgroupDevices(t *testing.T) {
+	nomad := e2eutil.NomadClient(t)
+
+	e2eutil.WaitForLeader(t, nomad)
+	e2eutil.WaitForNodesReady(t, nomad, 1)
+
+	t.Run("testDevicesUsable", testDevicesUsable)
+}
+
+func testDevicesUsable(t *testing.T) {
+	nomad := e2eutil.NomadClient(t)
+
+	jobID := "cgroup-devices-" + uuid.Short()
+	jobIDs := []string{jobID}
+	t.Cleanup(e2eutil.CleanupJobsAndGC(t, &jobIDs))
+
+	// start job
+	allocs := e2eutil.RegisterAndWaitForAllocs(t, nomad, "./input/cgroup_devices.hcl", jobID, "")
+	must.Len(t, 2, allocs)
+
+	// pick one to stop and one to verify
+	allocA := allocs[0].ID
+	allocB := allocs[1].ID
+
+	// verify devices are working
+	checkDev(t, allocA)
+	checkDev(t, allocB)
+
+	// stop the chosen alloc
+	_, err := e2eutil.Command("nomad", "alloc", "stop", "-detach", allocA)
+	must.NoError(t, err)
+	e2eutil.WaitForAllocStopped(t, nomad, allocA)
+
+	// verify device of remaining alloc
+	checkDev(t, allocB)
+}
+
+func checkDev(t *testing.T, allocID string) {
+	_, err := e2eutil.Command("nomad", "alloc", "exec", allocID, "dd", "if=/dev/zero", "of=/dev/null", "count=1")
+	must.NoError(t, err)
+}

--- a/e2e/isolation/input/cgroup_devices.hcl
+++ b/e2e/isolation/input/cgroup_devices.hcl
@@ -1,0 +1,41 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: MPL-2.0
+
+job "cgroup_devices" {
+  type = "service"
+
+  constraint {
+    attribute = "${attr.kernel.name}"
+    value     = "linux"
+  }
+
+  group "group1" {
+
+    task "task1" {
+      driver = "raw_exec"
+      config {
+        command = "/bin/sleep"
+        args    = ["infinity"]
+      }
+      resources {
+        cpu    = 50
+        memory = 50
+      }
+    }
+  }
+
+  group "group2" {
+
+    task "task2" {
+      driver = "raw_exec"
+      config {
+        command = "/bin/sleep"
+        args    = ["infinity"]
+      }
+      resources {
+        cpu    = 50
+        memory = 50
+      }
+    }
+  }
+}

